### DISCUSSION TITLE
halos and interpolation

### DISCUSSION THREAD

**Discussion by pjrusello** [CliMA/Oceananigans.jl#2689] [Computational science!] [answered]
I have a couple of questions related to interpolation of results near boundaries and the effect of halo regions.

If I use Oceananigans's `interpolate` method for Fields, it fails if the specified physical location is too close to a boundary with a bounds error since `interpoalte` tries to index outside the array bounds. `interpoalte` also fails if the interpolation location is equal to one of the center or face coordinates (i.e., no interpolation would be needed) since `interpolate` is still trying to do a linear interpolation rather than falling back to a table lookup (I think).

For the case of interpolating near boundaries though, I was wondering if it is advisable to use the halo region data for interpolation?

From what I understand the halos are there for calculations at the boundaries to maintain higher order accuracy throughout the domain. Is there any reason not to utilize them for lower order linear interpolation to align velocities and scalars in post-processing?

**Comment by simone-silvestri** [comment, marked as answer]
Hello! 

If you interpolate near the boundaries halo points should be used to interpolate. Take care that in case of `Bounded` domains only one  halo point is filled (since we cannot preserve high order in bounded domains).

The interpolation on a `Face` and `Center` is, indeed, still doing a linear interpolation that  but that should return the correct value regardless.

Can you provide an example for what setting make the interpolation fails? 
If I run this script the interpolation I am not able to reproduce the error

```
grid = RectilinearGrid(size = (5, 1, 1), x = (0, 5), z = (0, 1), y = (0, 1), topology = (Bounded, Bounded, Bounded))
field = CenterField(grid)
set!(field, (x, y, z) -> x)
fill_halo_regions!(field)
```
```
julia> interpolate(field, 1.0, 0.5, 0.5)
0.9999999999999998

julia> interpolate(field, 2.0, 0.5, 0.5)
2.0

julia> interpolate(field, 0.0, 0.5, 0.5)
0.0

julia> interpolate(field, 0.1, 0.5, 0.5)
0.10000000000000014
```

**Reply by glwagner** [reply]
Also in the future the halo points may not be filled, so interpolation is generally wrong outside the domain.

**Reply by pjrusello** [reply]
I tested this a bit more and I think the error I was encountering is entirely related to reading a `FieldTimeSeries` from a file. I tested with the Langmuir turbulence example at the end of the run using:  
```
u, v, w = model.velocities
interpolate( u, 128, 10, -10 )
```
Worked fine since the halo data is in the u array, similar to the example Simone provided above.

After loading the field output file with `u = FieldTimeSeries( "langmuir_turbulence_fields.jld2", "u", times = 0 )` the same interpolation will fail (I specified `with_halos = true` for the turbulence fields file and verified it's there by inspecting the file in HDFView). The u array size is 32x32x32x1 so the halo data isn't loaded and the above interpolation fails with a bounds error.

If I instead try `interpolate( u, 124, 38, -33 )` which should be the last face value and a center location in y and z in the domain, I still get a bounds error. Using `interpolate( u, 123.99, 38, -33 )`  works fine though since `interpolate` is using the last face value for the interpolation. I don't think this is a bug, just a limitation of how `interpolate` is written. 

Is there an option to force `FieldTimeSeries` to load the halo data?

Last thing, the above is an admittedly minor issue and I just want to say thanks for providing Oceananigans and making it so accessible.

**Reply by glwagner** [reply]
In this test, it looks like `FieldTimeSeries` does load halo data:

```julia
using Oceananigans
grid = RectilinearGrid(size=(4, 6, 8), extent=(1, 1, 1))
model = NonhydrostaticModel(; grid)
simulation = Simulation(model, Δt=1.0, stop_iteration=3)
simulation.output_writers[:out] = JLD2OutputWriter(model, model.velocities,
                                                   schedule = IterationInterval(1),
                                                   with_halos = true,
                                                   overwrite_existing = true,
                                                   filename = "test.jld2")

run!(simulation)

ut = FieldTimeSeries("test.jld2", "u")

@show size(ut)
@show size(ut.data)
```

gives me

```julia
julia> include("halo_test.jl")
[ Info: Initializing simulation...
[ Info:     ... simulation initialization complete (3.051 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (1.447 ms).
[ Info: Simulation is stopping. Model iteration 3 has hit or exceeded simulation stop iteration 3.
size(ut) = (4, 6, 8, 4)
size(ut.data) = (10, 12, 14, 4)
```

**Reply by glwagner** [reply]
> Last thing, the above is an admittedly minor issue and I just want to say thanks for providing Oceananigans and making it so accessible.

Of course! Thanks for helping to make it better!

**Reply by pjrusello** [reply]
Interesting. It does look like the halo data is loaded for the test I provided above
```
julia> u = FieldTimeSeries( "./WaveTransmittedTurbulence\\langmuir_turbulence_fields.jld2", "u" )
32×32×32×49 FieldTimeSeries{InMemory} located at (Face, Center, Center) on CPU
├── grid: 32×32×32 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── indices: (Colon(), Colon(), Colon())
└── data: 38×38×38×49 OffsetArray(::Array{Float64, 4}, -2:35, -2:35, -2:35, 1:49) with eltype Float64 with indices -2:35×-2:35×-2:35×1:49
    └── max=0.0872332, min=-0.0187338, mean=0.00349659
```
The only difference I can spot between this and the `u, v, w = model.velocities` version is the one loaded from the file is a `FieldTimeSeries` with an extra trailing dimension for time, while the one obtained from the model is just a `Field`. Ignore differences in array contents since these are not from the same timestep.

```
julia> u
32×32×32 Field{Face, Center, Center} on RectilinearGrid on CPU
├── grid: 32×32×32 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: Flux, immersed: ZeroFlux
└── data: 38×38×38 OffsetArray(::Array{Float64, 3}, -2:35, -2:35, -2:35) with eltype Float64 with indices -2:35×-2:35×-2:35
    └── max=0.0864187, min=-0.0188865, mean=0.00576265

julia> u_loaded
32×32×32×1 FieldTimeSeries{InMemory} located at (Face, Center, Center) on CPU
├── grid: 32×32×32 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── indices: (Colon(), Colon(), Colon())
└── data: 38×38×38×1 OffsetArray(::Array{Float64, 4}, -2:35, -2:35, -2:35, 1:1) with eltype Float64 with indices -2:35×-2:35×-2:35×1:1
    └── max=0.078604, min=-0.0133781, mean=0.00576279
```

Is there a way to convert a `FieldTimeSeries` to just a `Field`? Or should `interpolate` not care about that?

**Reply by glwagner** [reply]
You can obtain a `Field` by indexing into `FieldTimeSeries` at a particular time index:

```julia
u_time_series = FieldTimeSeries(langmuir_turbulence_fields.jld2", "u" )
u1 = u_time_series[1]
```

What are you trying to do?

We should certainly add some documentation for `FieldTimeSeries`...

**Reply by pjrusello** [reply]
I wanted to see if the interpolation issue is with how `Field` and `FieldTimeSeries` are handled by `interpolate`. I suspect this is by design, but this behavior isn't what I expected (u_loaded is the `FieldTimeSeries` above, u is the `Field`):
```
interpolate( u_loaded[ 1 ], 124, 38, -33 )
-0.0032558434177190065

julia> interpolate( u, 124, 38, -33 )
-0.0033981014731492613

julia> interpolate( u_loaded, 124, 38, -33 )
ERROR: BoundsError: attempt to access 32×32×32×1 FieldTimeSeries...
```
I'll chalk this up mostly to me not understanding the difference between a `Field` and a `FieldTimeSeries` (or noticing until now I was dealing with two different object types).

Ultimately, I was just trying to figure out the best way to interpolate u and v onto a common grid and was checking to see how Oceananigans or the examples handle this. Which led me down the rabbit hole of seeing what happens at boundaries and how to deal with halo regions and finding the above behavior.

**Reply by glwagner** [reply]
Ah! `interpolate` wasn't designed to work with `FieldTimeSeries`! This is an interesting application though.

`interpolate` leaves something to be desired, but it depends on what you are trying to use it for. If you would just like `u` and `v` on the same grid, then you can probably use `interpolate`. However, you can also directly output `u` and `v` on the tracer grid; for example:

```julia
uc = @at (Center, Center, Center) u
vc = @at (Center, Center, Center) v
```

and then pass `uc` and `vc` to an output writer.

But often you don't actually want `u` and `v` on the same grid. What is your application?

**Reply by pjrusello** [reply]
Cool. That is a super handy way to shift data output... Internally what's happening with the `@at` call? I would guess an interpolation, but is it higher order?

Beyond just a general interest in working with models more (my background is primarily experimental work), in this case we needed some realistic ocean surface velocity data to test a remote sensing model. Nothing that necessarily requires a high-fidelity solution, I'm mostly using it as an excuse to work with numerical models more.

**Reply by glwagner** [reply]
Ok, that sounds like an appropriate application!

If you inspect `uc` and `vc` you'll find that they are `UnaryOperation`, which is a type of "abstract operation". Abstract operations _represent_ operations (so they are "lazy"; we don't interpolate anything until we evaluate the operation, which typically occurs right before writing the data to disk). It's similar to, for example, something like `u2 = u^2`.
